File: example/src/main/java/com/seamlesspay/demo/MainActivity.java
/**
 * Copyright (c) Seamless Payments, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.seamlesspay.demo;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.Switch;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import com.google.android.material.tabs.TabLayout;
import com.seamlesspay.api.error.ApiError;
import com.seamlesspay.api.exceptions.InvalidArgumentException;
import com.seamlesspay.api.models.ClientConfiguration;
import com.seamlesspay.example.R;
import com.seamlesspay.ui.common.CardFormValidListener;
import com.seamlesspay.ui.common.PaymentCallback;
import com.seamlesspay.ui.common.TokenizeCallback;
import com.seamlesspay.ui.models.ChargeRequest;
import com.seamlesspay.ui.models.DisplayConfiguration;
import com.seamlesspay.ui.models.FieldConfiguration;
import com.seamlesspay.ui.models.FieldOptions;
import com.seamlesspay.ui.models.PaymentResponse;
import com.seamlesspay.ui.models.RefundRequest;
import com.seamlesspay.ui.models.TokenizeResponse;
import com.seamlesspay.ui.models.style.ColorPalette;
import com.seamlesspay.ui.models.style.Colors;
import com.seamlesspay.ui.models.style.ElevationLevel;
import com.seamlesspay.ui.models.style.IconSet;
import com.seamlesspay.ui.models.style.Shadow;
import com.seamlesspay.ui.models.style.Shapes;
import com.seamlesspay.ui.models.style.StyleOptions;
import com.seamlesspay.ui.models.style.ThemeColors;
import com.seamlesspay.ui.models.style.Typography;
import com.seamlesspay.ui.paymentinputs.direct.MultiLineCardForm;
import com.seamlesspay.ui.paymentinputs.direct.SingleLineCardForm;

public class MainActivity extends AppCompatActivity {
	SingleLineCardForm mSingleLineCardForm;
	MultiLineCardForm mMultiLineCardForm;
	TextView mInfoView;
	ProgressBar mProgressBar;
	TabLayout mTabLayout;
	boolean isMultiLine = false;

	private final TokenizeCallback tokenizeCallback = new TokenizeCallback() {
		@Override
		public void success(@NonNull TokenizeResponse tokenResponse) {
			mInfoView.setText(tokenResponse.toString());
			mProgressBar.setVisibility(View.GONE);
		}

		@SuppressLint("SetTextI18n")
		@Override
		public void failure(@Nullable ApiError apiError) {
			if (apiError != null) {
				mInfoView.setText("Error\n" + apiError.getStatusDescription());
			}
			mProgressBar.setVisibility(View.GONE);
		}
	};

	private final PaymentCallback paymentCallback = new PaymentCallback() {
		@Override
		public void success(@NonNull PaymentResponse paymentResponse) {
			mInfoView.setText(paymentResponse.toString());
			mProgressBar.setVisibility(View.GONE);
		}

		@SuppressLint("SetTextI18n")
		@Override
		public void failure(@Nullable ApiError apiError) {
			if (apiError != null) {
				mInfoView.setText("Error\n" + apiError.getStatusDescription());
			}
			mProgressBar.setVisibility(View.GONE);
		}
	};

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		setContentView(R.layout.activity_main);

		Button payButton = findViewById(R.id.payButton);
		Button tokenizeButton = findViewById(R.id.tokenizeButton);
		Button refundButton = findViewById(R.id.refundButton);
		mProgressBar = findViewById(R.id.progress);
		mTabLayout = findViewById(R.id.tabLayout);
		mInfoView = findViewById(R.id.infoView);
		mSingleLineCardForm = findViewById(R.id.cardSingleLine);
		mMultiLineCardForm = findViewById(R.id.cardMultiLine);
		try {
			StyleOptions styleOptions = new StyleOptions(
					new Colors(
							new ColorPalette(new ThemeColors(Color.RED, Color.BLUE, Color.MAGENTA)),
							new ColorPalette(new ThemeColors(Color.RED, Color.BLUE, Color.MAGENTA))
					),
					new Shapes(50F, new Shadow(ElevationLevel.Level1)),
					new Typography(com.seamlesspay.R.font.roboto_regular, 1f),
					IconSet.DARK
			);
			ClientConfiguration clientConfiguration =
					ClientConfiguration.fromKeys("staging", "pk_XXXXXXXXXXXXXXXXXXXXXXXXXX", "MRT_XXXXXXXXXXXXXXXXXXXXXXXXXX");
			FieldOptions option = new FieldOptions(new FieldConfiguration(DisplayConfiguration.OPTIONAL),
					new FieldConfiguration(DisplayConfiguration.REQUIRED));
			mSingleLineCardForm.init(clientConfiguration, option);
			mMultiLineCardForm.init(clientConfiguration, null, null);
		} catch (InvalidArgumentException e) {
			mInfoView.setText(e.getMessage());
		}

		mSingleLineCardForm.setOnCardFormValidListener(
				valid -> Log.v("MainActivity", "SingleLine valid: " + valid));

		setUpTabLayout();

		refundButton.setOnClickListener(v -> {
			// Code here executes on main thread after user presses button
			mSingleLineCardForm.clearFocus();
			mMultiLineCardForm.clearFocus();
			mInfoView.setText("");
			mProgressBar.setVisibility(View.VISIBLE);

			RefundRequest refundRequest = new RefundRequest(100);
			if (isMultiLine) {
				mMultiLineCardForm.refund(refundRequest, paymentCallback);
			} else {
				mSingleLineCardForm.refund(refundRequest, paymentCallback);
			}
		});

		tokenizeButton.setOnClickListener(v -> {
			// Code here executes on main thread after user presses button
			mSingleLineCardForm.clearFocus();
			mMultiLineCardForm.clearFocus();
			mInfoView.setText("");
			mProgressBar.setVisibility(View.VISIBLE);

			if (isMultiLine) {
				mMultiLineCardForm.tokenize(tokenizeCallback);
			} else {
				mSingleLineCardForm.tokenize(tokenizeCallback);
			}
		});

		payButton.setOnClickListener(v -> {
			// Code here executes on main thread after user presses button
			mSingleLineCardForm.clearFocus();
			mMultiLineCardForm.clearFocus();
			mInfoView.setText("");
			mProgressBar.setVisibility(View.VISIBLE);

			ChargeRequest chargeRequest = new ChargeRequest(100, true);
			if (isMultiLine) {
				mMultiLineCardForm.charge(chargeRequest, paymentCallback);
			} else {
				mSingleLineCardForm.charge(chargeRequest, paymentCallback);
			}
		});
	}

	private void setUpTabLayout() {
		mTabLayout.addOnTabSelectedListener(new TabLayout.BaseOnTabSelectedListener() {
			@Override
			public void onTabSelected(TabLayout.Tab tab) {
				if (tab.getPosition() == 1) {
					isMultiLine = true;
					mSingleLineCardForm.setVisibility(View.GONE);
					mMultiLineCardForm.setVisibility(View.VISIBLE);
				} else {
					isMultiLine = false;
					mSingleLineCardForm.setVisibility(View.VISIBLE);
					mMultiLineCardForm.setVisibility(View.GONE);
				}
			}

			@Override
			public void onTabUnselected(TabLayout.Tab tab) {

			}

			@Override
			public void onTabReselected(TabLayout.Tab tab) {

			}
		});
	}
}

File: example/src/main/java/com/seamlesspay/demo/MainActivity.kt
package com.seamlesspay.demo

import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.seamlesspay.api.error.ApiError
import com.seamlesspay.api.exceptions.InvalidArgumentException
import com.seamlesspay.api.models.ClientConfiguration
import com.seamlesspay.example.R
import com.seamlesspay.ui.common.PaymentCallback
import com.seamlesspay.ui.common.TokenizeCallback
import com.seamlesspay.ui.models.ChargeRequest
import com.seamlesspay.ui.models.DisplayConfiguration
import com.seamlesspay.ui.models.FieldConfiguration
import com.seamlesspay.ui.models.FieldOptions
import com.seamlesspay.ui.models.PaymentResponse
import com.seamlesspay.ui.models.RefundRequest
import com.seamlesspay.ui.models.TokenizeResponse
import com.seamlesspay.ui.models.style.ColorPalette
import com.seamlesspay.ui.models.style.Colors
import com.seamlesspay.ui.models.style.ElevationLevel
import com.seamlesspay.ui.models.style.FieldColors
import com.seamlesspay.ui.models.style.Shadow
import com.seamlesspay.ui.models.style.Shapes
import com.seamlesspay.ui.models.style.StyleOptions
import com.seamlesspay.ui.models.style.ThemeColors
import com.seamlesspay.ui.models.style.Typography
import com.seamlesspay.ui.paymentinputs.direct.CardForm

class MainActivity : AppCompatActivity() {

  // Main UI Components
  private lateinit var cardForm: CardForm
  private lateinit var infoView: TextView
  private lateinit var progressBar: ProgressBar


  // Tokenize and Payment callbacks
  private val tokenizeCallback: TokenizeCallback = object : TokenizeCallback {
    override fun success(tokenizeResponse: TokenizeResponse) {
      infoView.text = tokenizeResponse.toString()
      progressBar.visibility = View.GONE
    }

    @SuppressLint("SetTextI18n")
    override fun failure(apiError: ApiError?) {
      apiError?.let {
        infoView.text = "Error ${it.statusDescription}"
      }
      progressBar.visibility = View.GONE
    }
  }

  private val paymentCallback: PaymentCallback = object : PaymentCallback {
    override fun success(paymentResponse: PaymentResponse) {
      infoView.text = paymentResponse.toString()
      progressBar.visibility = View.GONE
    }

    @SuppressLint("SetTextI18n")
    override fun failure(apiError: ApiError?) {
      apiError?.let {
        infoView.text = "Error ${it.statusDescription}"
      }
      progressBar.visibility = View.GONE
    }
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    val payButton = findViewById<Button>(R.id.payButton)
    val tokenizeButton = findViewById<Button>(R.id.tokenizeButton)
    val refundButton = findViewById<Button>(R.id.refundButton)

    progressBar = findViewById(R.id.progress)
    infoView = findViewById(R.id.infoView)
    cardForm = findViewById(R.id.cardForm)

    initStyles()

    cardForm.setOnCardFormValidListener { valid: Boolean ->
      Log.v(
        "MainActivity",
        "CardForm valid: $valid"
      )
    }

    refundButton.setOnClickListener { _ ->
      // Code here executes on main thread after user presses button
      cardForm.clearFocus()
      infoView.text = ""
      progressBar.visibility = View.VISIBLE
      val refundRequest = RefundRequest(100)
      cardForm.refund(refundRequest, paymentCallback)
    }

    tokenizeButton.setOnClickListener { _ ->
      // Code here executes on main thread after user presses button
      cardForm.clearFocus()
      infoView.text = ""
      progressBar.visibility = View.VISIBLE
      cardForm.tokenize(tokenizeCallback)
    }

    payButton.setOnClickListener { _ ->
      // Code here executes on main thread after user presses button
      cardForm.clearFocus()
      infoView.text = ""
      progressBar.visibility = View.VISIBLE
      val chargeRequest = ChargeRequest(100, true)
      cardForm.charge(chargeRequest, paymentCallback)
    }
  }

  private fun initStyles() {
    try {
      //Set up custom styles for CardForm
      val styleOptions = StyleOptions(
        Colors(
          ColorPalette(
            ThemeColors(Color.RED, Color.BLUE, Color.MAGENTA),
            FieldColors(
              null,
              null,
              null,
              null,
              null,
              null
            ), Color.BLUE,
            Color.RED
          ),
          ColorPalette(
            ThemeColors(Color.RED, Color.BLUE, Color.MAGENTA), null, null,
            null
          )
        ), Shapes(50f, Shadow(ElevationLevel.Level2)),
        Typography(com.seamlesspay.R.font.roboto_regular, 1f), null
      )

      //Set up client configuration
      val clientConfiguration = ClientConfiguration.fromKeys(
        "staging",
        "pk_XXXXXXXXXXXXXXXXXXXXXXXXXX",
        "MRT_XXXXXXXXXXXXXXXXXXXXXXXXXX"
      )

      //Set up field options
      val fieldOptions = FieldOptions(
        FieldConfiguration(DisplayConfiguration.OPTIONAL),
        FieldConfiguration(DisplayConfiguration.REQUIRED)
      )

      //Init card form with you custom configurations
      cardForm.init(clientConfiguration, fieldOptions)
    } catch (e: InvalidArgumentException) {
      infoView.text = e.message
    }
  }
}